// @typespec/compiler/dist/src/core/types.js
export var ResolutionResultFlags;
(function (ResolutionResultFlags) {
    ResolutionResultFlags[ResolutionResultFlags["None"] = 0] = "None";
    ResolutionResultFlags[ResolutionResultFlags["Resolved"] = 2] = "Resolved";
    ResolutionResultFlags[ResolutionResultFlags["Unknown"] = 4] = "Unknown";
    ResolutionResultFlags[ResolutionResultFlags["Ambiguous"] = 8] = "Ambiguous";
    ResolutionResultFlags[ResolutionResultFlags["NotFound"] = 16] = "NotFound";
    ResolutionResultFlags[ResolutionResultFlags["ResolutionFailed"] = 28] = "ResolutionFailed";
})(ResolutionResultFlags || (ResolutionResultFlags = {}));
/**
 * AST types
 */
export var SyntaxKind;
(function (SyntaxKind) {
    SyntaxKind[SyntaxKind["TypeSpecScript"] = 0] = "TypeSpecScript";
    SyntaxKind[SyntaxKind["JsSourceFile"] = 1] = "JsSourceFile";
    SyntaxKind[SyntaxKind["ImportStatement"] = 2] = "ImportStatement";
    SyntaxKind[SyntaxKind["Identifier"] = 3] = "Identifier";
    SyntaxKind[SyntaxKind["AugmentDecoratorStatement"] = 4] = "AugmentDecoratorStatement";
    SyntaxKind[SyntaxKind["DecoratorExpression"] = 5] = "DecoratorExpression";
    SyntaxKind[SyntaxKind["DirectiveExpression"] = 6] = "DirectiveExpression";
    SyntaxKind[SyntaxKind["MemberExpression"] = 7] = "MemberExpression";
    SyntaxKind[SyntaxKind["NamespaceStatement"] = 8] = "NamespaceStatement";
    SyntaxKind[SyntaxKind["UsingStatement"] = 9] = "UsingStatement";
    SyntaxKind[SyntaxKind["OperationStatement"] = 10] = "OperationStatement";
    SyntaxKind[SyntaxKind["OperationSignatureDeclaration"] = 11] = "OperationSignatureDeclaration";
    SyntaxKind[SyntaxKind["OperationSignatureReference"] = 12] = "OperationSignatureReference";
    SyntaxKind[SyntaxKind["ModelStatement"] = 13] = "ModelStatement";
    SyntaxKind[SyntaxKind["ModelExpression"] = 14] = "ModelExpression";
    SyntaxKind[SyntaxKind["ModelProperty"] = 15] = "ModelProperty";
    SyntaxKind[SyntaxKind["ModelSpreadProperty"] = 16] = "ModelSpreadProperty";
    SyntaxKind[SyntaxKind["ScalarStatement"] = 17] = "ScalarStatement";
    SyntaxKind[SyntaxKind["InterfaceStatement"] = 18] = "InterfaceStatement";
    SyntaxKind[SyntaxKind["UnionStatement"] = 19] = "UnionStatement";
    SyntaxKind[SyntaxKind["UnionVariant"] = 20] = "UnionVariant";
    SyntaxKind[SyntaxKind["EnumStatement"] = 21] = "EnumStatement";
    SyntaxKind[SyntaxKind["EnumMember"] = 22] = "EnumMember";
    SyntaxKind[SyntaxKind["EnumSpreadMember"] = 23] = "EnumSpreadMember";
    SyntaxKind[SyntaxKind["AliasStatement"] = 24] = "AliasStatement";
    SyntaxKind[SyntaxKind["DecoratorDeclarationStatement"] = 25] = "DecoratorDeclarationStatement";
    SyntaxKind[SyntaxKind["FunctionDeclarationStatement"] = 26] = "FunctionDeclarationStatement";
    SyntaxKind[SyntaxKind["FunctionParameter"] = 27] = "FunctionParameter";
    SyntaxKind[SyntaxKind["UnionExpression"] = 28] = "UnionExpression";
    SyntaxKind[SyntaxKind["IntersectionExpression"] = 29] = "IntersectionExpression";
    SyntaxKind[SyntaxKind["TupleExpression"] = 30] = "TupleExpression";
    SyntaxKind[SyntaxKind["ArrayExpression"] = 31] = "ArrayExpression";
    SyntaxKind[SyntaxKind["StringLiteral"] = 32] = "StringLiteral";
    SyntaxKind[SyntaxKind["NumericLiteral"] = 33] = "NumericLiteral";
    SyntaxKind[SyntaxKind["BooleanLiteral"] = 34] = "BooleanLiteral";
    SyntaxKind[SyntaxKind["StringTemplateExpression"] = 35] = "StringTemplateExpression";
    SyntaxKind[SyntaxKind["StringTemplateHead"] = 36] = "StringTemplateHead";
    SyntaxKind[SyntaxKind["StringTemplateMiddle"] = 37] = "StringTemplateMiddle";
    SyntaxKind[SyntaxKind["StringTemplateTail"] = 38] = "StringTemplateTail";
    SyntaxKind[SyntaxKind["StringTemplateSpan"] = 39] = "StringTemplateSpan";
    SyntaxKind[SyntaxKind["ExternKeyword"] = 40] = "ExternKeyword";
    SyntaxKind[SyntaxKind["VoidKeyword"] = 41] = "VoidKeyword";
    SyntaxKind[SyntaxKind["NeverKeyword"] = 42] = "NeverKeyword";
    SyntaxKind[SyntaxKind["UnknownKeyword"] = 43] = "UnknownKeyword";
    SyntaxKind[SyntaxKind["ValueOfExpression"] = 44] = "ValueOfExpression";
    SyntaxKind[SyntaxKind["TypeReference"] = 45] = "TypeReference";
    SyntaxKind[SyntaxKind["TemplateParameterDeclaration"] = 46] = "TemplateParameterDeclaration";
    SyntaxKind[SyntaxKind["EmptyStatement"] = 47] = "EmptyStatement";
    SyntaxKind[SyntaxKind["InvalidStatement"] = 48] = "InvalidStatement";
    SyntaxKind[SyntaxKind["LineComment"] = 49] = "LineComment";
    SyntaxKind[SyntaxKind["BlockComment"] = 50] = "BlockComment";
    SyntaxKind[SyntaxKind["Doc"] = 51] = "Doc";
    SyntaxKind[SyntaxKind["DocText"] = 52] = "DocText";
    SyntaxKind[SyntaxKind["DocParamTag"] = 53] = "DocParamTag";
    SyntaxKind[SyntaxKind["DocPropTag"] = 54] = "DocPropTag";
    SyntaxKind[SyntaxKind["DocReturnsTag"] = 55] = "DocReturnsTag";
    SyntaxKind[SyntaxKind["DocErrorsTag"] = 56] = "DocErrorsTag";
    SyntaxKind[SyntaxKind["DocTemplateTag"] = 57] = "DocTemplateTag";
    SyntaxKind[SyntaxKind["DocUnknownTag"] = 58] = "DocUnknownTag";
    SyntaxKind[SyntaxKind["Return"] = 59] = "Return";
    SyntaxKind[SyntaxKind["JsNamespaceDeclaration"] = 60] = "JsNamespaceDeclaration";
    SyntaxKind[SyntaxKind["TemplateArgument"] = 61] = "TemplateArgument";
    SyntaxKind[SyntaxKind["TypeOfExpression"] = 62] = "TypeOfExpression";
    SyntaxKind[SyntaxKind["ObjectLiteral"] = 63] = "ObjectLiteral";
    SyntaxKind[SyntaxKind["ObjectLiteralProperty"] = 64] = "ObjectLiteralProperty";
    SyntaxKind[SyntaxKind["ObjectLiteralSpreadProperty"] = 65] = "ObjectLiteralSpreadProperty";
    SyntaxKind[SyntaxKind["ArrayLiteral"] = 66] = "ArrayLiteral";
    SyntaxKind[SyntaxKind["ConstStatement"] = 67] = "ConstStatement";
    SyntaxKind[SyntaxKind["CallExpression"] = 68] = "CallExpression";
    SyntaxKind[SyntaxKind["ScalarConstructor"] = 69] = "ScalarConstructor";
})(SyntaxKind || (SyntaxKind = {}));
export var IdentifierKind;
(function (IdentifierKind) {
    IdentifierKind[IdentifierKind["TypeReference"] = 0] = "TypeReference";
    IdentifierKind[IdentifierKind["TemplateArgument"] = 1] = "TemplateArgument";
    IdentifierKind[IdentifierKind["Decorator"] = 2] = "Decorator";
    IdentifierKind[IdentifierKind["Function"] = 3] = "Function";
    IdentifierKind[IdentifierKind["Using"] = 4] = "Using";
    IdentifierKind[IdentifierKind["Declaration"] = 5] = "Declaration";
    IdentifierKind[IdentifierKind["ModelExpressionProperty"] = 6] = "ModelExpressionProperty";
    IdentifierKind[IdentifierKind["ModelStatementProperty"] = 7] = "ModelStatementProperty";
    IdentifierKind[IdentifierKind["ObjectLiteralProperty"] = 8] = "ObjectLiteralProperty";
    IdentifierKind[IdentifierKind["Other"] = 9] = "Other";
})(IdentifierKind || (IdentifierKind = {}));
/** Used to explicitly specify that a diagnostic has no target. */
export const NoTarget = Symbol.for("NoTarget");
export var ListenerFlow;
(function (ListenerFlow) {
    /**
     * Do not navigate any containing or referenced type.
     */
    ListenerFlow[ListenerFlow["NoRecursion"] = 1] = "NoRecursion";
})(ListenerFlow || (ListenerFlow = {}));
