// @azure-tools/typespec-azure-core/dist/src/tsp-index.js
import { $armResourceIdentifierConfig, $defaultFinalStateVia, $embeddingVector, $ensureResourceType, $ensureVerb, $finalLocation, $finalOperation, $fixed, $items, $lroCanceled, $lroErrorResult, $lroFailed, $lroResult, $lroStatus, $lroSucceeded, $needsRoute, $nextPageOperation, $omitKeyProperties, $operationLink, $pagedResult, $pollingLocation, $pollingOperation, $pollingOperationParameter, $requestParameter, $responseProperty, $spreadCustomParameters, $spreadCustomResponseProperties, $useFinalStateVia, parameterizedNextLinkConfigDecorator, } from "./decorators.js";
import { $addTraitProperties, $applyTraitOverride, $ensureAllHeaderParams, $ensureAllQueryParams, $ensureTraitsPresent, $trait, $traitAdded, $traitContext, $traitLocation, $traitSource, } from "./traits.js";
export { $lib } from "./lib.js";
export { $onValidate } from "./validate.js";
/** @internal */
export const $decorators = {
    "Azure.Core": {
        lroStatus: $lroStatus,
        finalLocation: $finalLocation,
        pollingLocation: $pollingLocation,
        pagedResult: $pagedResult,
        items: $items,
        fixed: $fixed,
        lroSucceeded: $lroSucceeded,
        lroCanceled: $lroCanceled,
        lroFailed: $lroFailed,
        lroResult: $lroResult,
        lroErrorResult: $lroErrorResult,
        operationLink: $operationLink,
        pollingOperationParameter: $pollingOperationParameter,
        pollingOperation: $pollingOperation,
        finalOperation: $finalOperation,
        useFinalStateVia: $useFinalStateVia,
        nextPageOperation: $nextPageOperation,
    },
    "Azure.Core.Foundations": {
        omitKeyProperties: $omitKeyProperties,
        requestParameter: $requestParameter,
        responseProperty: $responseProperty,
    },
    "Azure.Core.Foundations.Private": {
        spreadCustomParameters: $spreadCustomParameters,
        spreadCustomResponseProperties: $spreadCustomResponseProperties,
        ensureResourceType: $ensureResourceType,
        embeddingVector: $embeddingVector,
        armResourceIdentifierConfig: $armResourceIdentifierConfig,
        needsRoute: $needsRoute,
        ensureVerb: $ensureVerb,
        defaultFinalStateVia: $defaultFinalStateVia,
        parameterizedNextLinkConfig: parameterizedNextLinkConfigDecorator,
    },
    "Azure.Core.Traits": {
        trait: $trait,
        traitAdded: $traitAdded,
        traitContext: $traitContext,
        traitLocation: $traitLocation,
    },
    "Azure.Core.Traits.Private": {
        applyTraitOverride: $applyTraitOverride,
        ensureAllQueryParams: $ensureAllQueryParams,
        ensureAllHeaderParams: $ensureAllHeaderParams,
        addTraitProperties: $addTraitProperties,
        traitSource: $traitSource,
        ensureTraitsPresent: $ensureTraitsPresent,
    },
};
