// @azure-tools/typespec-azure-core/dist/src/rules/no-private-usage.js
import { createRule, paramMessage, } from "@typespec/compiler";
import { checkDecoratorsInDisallowedNamespace, checkReferenceInDisallowedNamespace, } from "./utils.js";
export const noPrivateUsage = createRule({
    name: "no-private-usage",
    description: "Verify that elements inside Private namespace are not referenced.",
    severity: "warning",
    url: "https://azure.github.io/typespec-azure/docs/libraries/azure-core/rules/no-private-usage",
    messages: {
        default: paramMessage `Referencing elements inside Private namespace "${"ns"}" is not allowed.`,
    },
    create(context) {
        function checkReference(origin, type, target) {
            return checkReferenceInDisallowedNamespace(context, origin, type, target, "Private");
        }
        function checkDecorators(type) {
            return checkDecoratorsInDisallowedNamespace(context, type, "Private");
        }
        return {
            model: (model) => {
                checkDecorators(model);
                model.baseModel && checkReference(model, model.baseModel, model);
            },
            modelProperty: (prop) => {
                checkDecorators(prop);
                checkReference(prop, prop.type, prop);
            },
            unionVariant: (variant) => {
                checkDecorators(variant);
                checkReference(variant, variant.type, variant);
            },
            operation: (type) => {
                checkDecorators(type);
            },
            interface: (type) => {
                checkDecorators(type);
            },
            enum: (type) => {
                checkDecorators(type);
            },
            union: (type) => {
                checkDecorators(type);
            },
        };
    },
});
